### DISCUSSION TITLE
Help for PEPPOL reporting

### DISCUSSION THREAD

**Discussion by dzagagnoni89** [phax/phase4#113] [Q&A] [answered]
Hello,
in my company we use **phase4** for PEPPOL both in reception and in sending. 

We are analyzing the specifications indicated in [this ](https://test-docs.peppol.eu/reporting/EUR-1.0.0-RC2/tsr/bis/) article to start implementing PEPPOL reporting and we have some questions:

1. In the sending phase, where do we find the "_**Timing of the exchange**_" field? When receiving it should be here --> AS4 field "_eb:UserMessage/eb:MessageInfo/eb:Timestamp_".
2. Is there a method or a function to easily extract the **_PEPPOL SeatID_** of the participants from their certificate?
3. Where do we find the **_Country Id_** of the Sender and Country Id of the Receiver fields?

Thanks in advance for the support!
Regards,
Davide Zagagnoni

**Comment by phax** [comment, marked as answer]
Hi Davide, thanks for reaching out :)

Please note, that I also started an Open Source implementation for Peppol Reporting based on the specifications you mentioned - it can be found at https://github.com/phax/peppol-reporting (under development).

Regarding your questions:
ad 1. In the first version of the BIS sent out for review, a "Guideline for Service Providers" was included - that is still available under https://openpeppol.atlassian.net/wiki/download/attachments/2832957455/Peppol%20Reporting%20-%20SP%20Guideline%20v1.0.pdf?api=v2
It states this on Timing:

![grafik](https://user-images.githubusercontent.com/196219/212058479-de8c4b13-e9ca-4e7f-901d-f58f57325eed.png)

ad 2. I already did this code in class `PeppolCertificateHelper` - you can extract it either from a Java `X509Certificate` or a Subject String.

ad 3. The Country codes of Sender and Receiver will be part of the SBDH. That is an update of the SBDH specification that requires publication and therefore these 2 fields are currently optional. See this section of the BIS you are referencing:

![grafik](https://user-images.githubusercontent.com/196219/212059058-af7ad118-6900-4dad-ad7d-401549ffbb27.png)

hth, Philip

**Reply by dzagagnoni89** [reply]
Hi Philip,
Thank you for your quick and clear response.
I will follow with great pleasure the project you are developing and I take this opportunity to thank you for the excellent work you have done with your other software.
See you soon.
Greetings,
Davide Zagagnoni
